hibernate-core/src/main/java/org/hibernate/query/sqm/tree/insert/SqmInsertValuesStatement.java

```diff
@@ -108,11 +114,29 @@ public <X> X accept(SemanticQueryWalker<X> walker) {
 		return walker.visitInsertValuesStatement( this );
 	}
 
+	@Override
+	public <U> Subquery<U> subquery(EntityType<U> type) {
+		throw new UnsupportedOperationException( "INSERT query cannot be sub-query" );
+	}
+
 	@Override
 	public JpaPredicate getRestriction() {
 		return null;
 	}
 
+	@Override
+	public Set<ParameterExpression<?>> getParameters() {
+		// At this level, the number of parameters may still be growing as
+		// nodes are added to the Criteria - so we re-calculate this every
+		// time.
+		//
+		// for a "finalized" set of parameters, use `#resolveParameters` instead
+		assert getQuerySource() == SqmQuerySource.CRITERIA;
+		return getSqmParameters().stream()
```

Comment:
I don't know what these mean.  You keep referring to some comments on "the other SqmStatement implementations" that as far as I can tell don't exist.

Reply:
IMO this `getParameters()` method implementation should move up to `org.hibernate.query.sqm.tree.AbstractSqmStatement`.

Also, every implementation of the `SqmStatement` interface implements `subquery(Class)` by returning a proper subquery object, so the `subquery(EntityType)` method should IMO not throw an exception. I also think we should move the implementation of the two `subquery` methods up to `AbstractSqmStatement` if possible to avoid duplication.